### PR TITLE
ZCS-916:Implement soap API for elsif and else in sieve script

### PR DESCRIPTION
**Pull request by shrikant-prasad**
Implemented soap2sieve and sieve2soap api for elsif and else rules in sieve script
Added Junit tests for sieve2soap changes. Junits for soap2sieve will be sent in separate PR.

### CODE REVIEW

File: soap/src/java/com/zimbra/soap/mail/type/NestedRule.java [hunk 1]
```diff
@@ -144,12 +152,48 @@ public void setChild(NestedRule nestedRule) {
         child = nestedRule;
     }
 
+    public void setElseRules(Iterable <NestedRule> elseRules) {
+        if(this.elseRules == null){
```
Comment (ykomiyam) [inline soap/src/java/com/zimbra/soap/mail/type/NestedRule.java]: add a space between "if" and "("

File: soap/src/java/com/zimbra/soap/mail/type/NestedRule.java [hunk 2]
```diff
@@ -144,12 +152,48 @@ public void setChild(NestedRule nestedRule) {
...
+            this.elseRules.clear();
+        }
+
+        if (elseRules != null) {
+            Iterables.addAll(this.elseRules,elseRules);
+        }
+    }
+
+    public void addElseRule(NestedRule elseRule) {
+        if(this.elseRules == null){
```
Comment (ykomiyam) [inline soap/src/java/com/zimbra/soap/mail/type/NestedRule.java]: ditto

File: soap/src/java/com/zimbra/soap/mail/type/NestedRule.java [hunk 3]
```diff
@@ -144,12 +152,48 @@ public void setChild(NestedRule nestedRule) {
...
+            this.elseRules = Lists.newArrayList();
+        }
+        this.elseRules.add(elseRule);
+    }
+
+    public void addElseRules(Iterable <NestedRule> elseRules) {
+        if(this.elseRules == null){
```
Comment (ykomiyam) [inline soap/src/java/com/zimbra/soap/mail/type/NestedRule.java]: ditto

File: store/src/java/com/zimbra/cs/filter/SoapToSieve.java [hunk 4]
```diff
@@ -206,6 +240,25 @@ private String handleNest(String baseIndents, NestedRule currentNestedRule) thro
...
+        if(currentNestedRule.getElseRules() != null && currentNestedRule.getElseRules().size() > 0){
+            if(isElse) {
+                // elseRule can't have nested elseRule
+                throw ServiceException.INVALID_REQUEST("elseRule can't have nested elseRule", null);
+            }
+            boolean elseFound = false;
+            for (NestedRule elseRule : currentNestedRule.getElseRules()) {
+                if (elseFound) {
+                    // elsif or else MUST only follow an if or elsif
+                    throw ServiceException.INVALID_REQUEST("Else or elsif can't follow else", null);
```
Comment (ykomiyam) [inline store/src/java/com/zimbra/cs/filter/SoapToSieve.java]: Perhaps the "Else or ..." may start with a lower-case letter as similar as the line 247?

File: store/src/java/com/zimbra/cs/filter/SoapToSieve.java [hunk 5]
```diff
@@ -143,38 +143,72 @@ private void handleRule(FilterRule rule) throws ServiceException {
...
+        handleElseRules(rule.getElseRules());
+    }
+
+    private void handleElseRules(List<NestedRule> elseRules) throws ServiceException {
+        boolean elseFound = false;
+        if(elseRules != null && elseRules.size() > 0){
+            for (NestedRule elseRule : elseRules) {
+                if (elseFound) {
+                    // elsif or else MUST only follow an if or elsif
+                    throw ServiceException.INVALID_REQUEST("Else or elsif can't follow else", null);
```
Comment (ykomiyam) [inline store/src/java/com/zimbra/cs/filter/SoapToSieve.java]: The "Else or ..." may start with a lower-case letter as "else or ..." ?

File: store/src/java/com/zimbra/cs/filter/SieveVisitor.java [hunk 6]
```diff
@@ -304,13 +307,24 @@ private void accept(Node parent, RuleProperties props) throws ServiceException {
...
+                }
                 visitIfControl(node,VisitPhase.begin,newProps);
-                accept(node, newProps);
+                if (SieveConstants.ELSE.equalsIgnoreCase(nodeName)) {
+                    newProps.condition = null;
+                    visitRule(node, VisitPhase.begin, newProps);
+                    accept(node, newProps);
+                    visitRule(node, VisitPhase.end, newProps);
+                } else {
+                   accept(node, newProps);
```
Comment (ykomiyam) [inline store/src/java/com/zimbra/cs/filter/SieveVisitor.java]: Please align the indent.

File: store/src/java-test/com/zimbra/cs/filter/GetFilterRulesForElseRulesTest.java [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+
+    @Test
+    public void testIfElseIf() throws Exception {
+        Account acct = Provisioning.getInstance().get(Key.AccountBy.name, "test@zimbra.com");
+        RuleManager.clearCachedRules(acct);
+        String ss = "require [\"fileinto\", \"log\"];\n";
+        ss += "if anyof header :is \"Subject\" \"important\" {\n";
+        ss += "fileinto \"FromP2\"; log \"Move message to FromP2 folder\"; }\n";
+        ss += "elsif anyof (header :contains \"subject\" [\"automation\", \"nunit\"]) \n";
+        ss += "{ redirect \"qa-automation@zqa-380.eng.zimbra.com\"; log \"Forward message to qa-automation DL\"; }\n";
```
Comment (ykomiyam) [inline store/src/java-test/com/zimbra/cs/filter/GetFilterRulesForElseRulesTest.java]: Why don't you redirect the message to "redirect@zimbra.com" or something?
Although we can use arbitrary account and domain name for the unit test, I think it is convention that we use "@zimbra.com" as a domain name, and more general name than "qa" or "automation" as an account name.

File: soap/src/java/com/zimbra/soap/mail/type/NestedRule.java (hunk 1 above)
Comment (shrikant-prasad) [inline soap/src/java/com/zimbra/soap/mail/type/NestedRule.java]: done at all occurrences.

File: store/src/java/com/zimbra/cs/filter/SoapToSieve.java (hunk 5 above)
Comment (shrikant-prasad) [inline store/src/java/com/zimbra/cs/filter/SoapToSieve.java]: done

File: store/src/java/com/zimbra/cs/filter/SoapToSieve.java (hunk 4 above)
Comment (shrikant-prasad) [inline store/src/java/com/zimbra/cs/filter/SoapToSieve.java]: done

File: store/src/java/com/zimbra/cs/filter/SieveVisitor.java (hunk 6 above)
Comment (shrikant-prasad) [inline store/src/java/com/zimbra/cs/filter/SieveVisitor.java]: done

File: store/src/java-test/com/zimbra/cs/filter/GetFilterRulesForElseRulesTest.java (hunk 7 above)
Comment (shrikant-prasad) [inline store/src/java-test/com/zimbra/cs/filter/GetFilterRulesForElseRulesTest.java]: done.

**Comment by chinmay15** [comment]
I agree with @rohan-ambasta here. We can remove NestedRule/ElseRule and use a single FilterRule class to handle this. If we extend FilterAction on FilterRule, we probably can solve this problem easily by putting nested rules in action list. It might not be doable directly, but it is worth to attempt. 

Another approach, as rohan has mentioned, is using tree instead of stack, where child and parent relationship can be handled easily and in better way, further it will be easy to understand.

**Comment by shrikant-prasad** [comment]
NestedRule and FilterRule classes existed before this solution and for elseRule no new class has been introduced. The scope and behaviour of if, nested if, elsif and else rules are different thats why different stacks are needed. If we want to handle all using a single tree, we can open separate spike bug for that and see if it can done or not. 

FilterAction extending FilterRule doesn't makes any sense as both these classes represent separate entities action and rule, respectively.

**Pull request closed without merging**